dioptas/widgets/plot_widgets/utils.py

```diff
@@ -43,3 +45,133 @@ def auto_level(hist_x: np.ndarray, hist_y: np.ndarray) -> tuple[float, float]:
     if len(hist_x[hist_x > 0]) > 0:
         min_level = max(min_level, np.nanmin(hist_x[hist_x > 0]))
     return min_level, max_level
+
+
+def _default_auto_level(data: np.ndarray) -> tuple[float, float]:
+    """Compute colormap range from data through histogram
+
+    :param data: Data from which to compute levels
+    :returns: (lower limit, upper limit)
+    """
+    counts, bin_edges = np.histogram(data, bins=3000)
+    left_edges = bin_edges[:-1]
+    left_edges = left_edges[np.where(counts > 0)]
+    counts = counts[np.where(counts > 0)]
+
+    hist_y_cumsum = np.cumsum(counts)
+    hist_y_sum = np.sum(counts)
+
+    max_ind = np.where(hist_y_cumsum < (0.996 * hist_y_sum))
+    min_level = np.mean(left_edges[:2])
+
+    if len(max_ind[0]):
+        max_level = left_edges[max_ind[0][-1]]
+    else:
+        max_level = 0.5 * np.max(left_edges)
+
+    if len(left_edges[left_edges > 0]) > 0:
+        min_level = max(min_level, np.nanmin(left_edges[left_edges > 0]))
+    return min_level, max_level
+
+
+def _minmax_auto_level(data: np.ndarray) -> tuple[float, float]:
+    """Returns min/max of the data
+
+    :param data: Data from which to compute levels
+    :returns: (min, max)
+    """
+    return float(np.min(data)), float(np.max(data))
+
+
+def _mean3std_auto_level(data: np.ndarray) -> tuple[float, float]:
+    """Returns mean+/-3std clipped to min/max of the data
+
+    :param data: Data from which to compute levels
+    :returns: (lower limit, upper limit)
+    """
+    mean = np.mean(data, dtype=np.float64)
+    std = np.std(data, dtype=np.float64)
+    minimum, maximum = _minmax_auto_level(data)
+    return max(mean - 3 * std, minimum), min(mean + 3 * std, maximum)
+
+
+def _percentile_auto_level(
+    data: np.ndarray, percentile: float = 1.0
+) -> tuple[float, float]:
+    """Returns data range corresponding to [precentile, 100-percentile]
+
+    :param data: Data from which to compute levels
+    :returns: (lower limit, upper limit)
+    """
+    if not 0.0 <= percentile < 50.0:
+        raise ValueError("Percentiles must be in the range [0, 50[")
+    lower, upper = np.percentile(data, [percentile, 100.0 - percentile])
+    return float(lower), float(upper)
+
+
+def detector_gap_mask(data: np.ndarray) -> np.ndarray:
+    """Probe detector gap value and returns mask of pixels not belonging to gaps.
+
+    :param data: 2D image array for which to compute the mask
+    :returns: Mask with True for valid pixels and False for mask
+    """
```

Comment:
I've tried a basic gap detection. This is useful with gap values that are large positive (negative) values.

Maybe the gap value can be guessed from the detector type?

Reply:
Since dioptas uses pyFAI's Detector, if there's a way to retrieve the selected Detector in the Calibration tab, it would be possible to leverage pyFAI's [Detector.dynamic_mask](https://github.com/silx-kit/pyFAI/blob/1fd17de90732493155959cfd03f893b1b3393f2d/pyFAI/detectors/_common.py#L861) to remove the bad pixels from the computation of the autoscale.

Reply:
Filter dummy values removed from this PR, I'll prepare a dedicated PR for it.